src/main/kotlin/org/rust/lang/core/psi/RsPsiManager.kt

```diff
@@ -0,0 +1,88 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.psi
+
+import com.intellij.openapi.components.ProjectComponent
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.ModificationTracker
+import com.intellij.openapi.util.SimpleModificationTracker
+import com.intellij.psi.*
+import com.intellij.psi.impl.PsiTreeChangeEventImpl
+import com.intellij.psi.util.PsiTreeUtil.getContextOfType
+import org.rust.lang.RsFileType
+import org.rust.lang.core.psi.ext.RsItemElement
+import org.rust.lang.core.psi.ext.RsModificationTrackerOwner
+
+interface RsPsiManager {
+    /**
+     * A project-global modification tracker that increments on each PSI change that can affect
+     * name resolution or type inference. It will be incremented with a change of most types of
+     * PSI element excluding function bodies (expressions and statements)
+     */
+    val rustStructureModificationTracker: ModificationTracker
+}
+
+class RsPsiManagerImpl(val project: Project) : ProjectComponent, RsPsiManager {
+
+    override val rustStructureModificationTracker = SimpleModificationTracker()
+
+    override fun projectOpened() {
+        PsiManager.getInstance(project).addPsiTreeChangeListener(CacheInvalidator())
+    }
+
+    inner class CacheInvalidator : PsiTreeChangeAdapter() {
+        override fun childRemoved(event: PsiTreeChangeEvent) = onPsiChange(event)
+        override fun childReplaced(event: PsiTreeChangeEvent) = onPsiChange(event)
+        override fun childAdded(event: PsiTreeChangeEvent) = onPsiChange(event)
+        override fun childrenChanged(event: PsiTreeChangeEvent) = onPsiChange(event)
+        override fun childMoved(event: PsiTreeChangeEvent) = onPsiChange(event)
+        override fun propertyChanged(event: PsiTreeChangeEvent) = onPsiChange(event)
+
+        private fun onPsiChange(event: PsiTreeChangeEvent) {
+            // `GenericChange` event means that "something changed in the file" and sends
+            // after all events for concrete PSI changes in a file.
+            // We handle more concrete events and so should ignore generic event
+            if (event is PsiTreeChangeEventImpl && event.isGenericChange) return
+
+            if(event.file?.fileType != RsFileType) return
+
+            val child = event.child
+            if (child is PsiComment || child is PsiWhiteSpace) return
+
+            updateModificationCount(child ?: event.parent)
+        }
+    }
+
+    private fun updateModificationCount(psi: PsiElement) {
+        // We find a nearest parent item or macro call (because macro call can produce items)
+        // If found item implements RsModificationTrackerOwner, we increment it's own
+        // modification counter. Otherwise we increment global modification counter.
+        //
+        // So, if we changing something inside a functions and it is not an item,
```

Comment:
Let's better write something like this:
"So, if something is changed inside a function except an item, we will only increment the function local modification counter."